mobile/src/foss/java/org/openhab/habdroid/ui/MapViewHelper.java

```diff
@@ -1,13 +1,255 @@
 import android.view.LayoutInflater;
 import android.view.ViewGroup;
 
+import androidx.core.content.ContextCompat;
+
+import org.openhab.habdroid.R;
 import org.openhab.habdroid.core.connection.Connection;
+import org.openhab.habdroid.model.Item;
+import org.openhab.habdroid.model.ParsedState;
+import org.openhab.habdroid.model.Widget;
+import org.openhab.habdroid.util.Util;
+import org.osmdroid.api.IMapController;
+import org.osmdroid.config.Configuration;
+import org.osmdroid.events.MapEventsReceiver;
+import org.osmdroid.tileprovider.tilesource.TileSourceFactory;
+import org.osmdroid.util.BoundingBox;
+import org.osmdroid.util.GeoPoint;
+import org.osmdroid.views.MapView;
+import org.osmdroid.views.overlay.CopyrightOverlay;
+import org.osmdroid.views.overlay.MapEventsOverlay;
+import org.osmdroid.views.overlay.Marker;
+
+import java.util.ArrayList;
+import java.util.Locale;
 
 public class MapViewHelper {
+    private static final String TAG = MapViewHelper.class.getSimpleName();
+
     public static WidgetAdapter.ViewHolder createViewHolder(LayoutInflater inflater,
             ViewGroup parent, Connection connection, WidgetAdapter.ColorMapper colorMapper) {
-        return new WidgetAdapter.GenericViewHolder(inflater, parent, connection, colorMapper);
+        Context context = inflater.getContext();
+        Configuration.getInstance().load(context,
+                PreferenceManager.getDefaultSharedPreferences(context));
+        return new OsmViewHolder(inflater, parent, connection, colorMapper);
+    }
+
+    private static class OsmViewHolder extends WidgetAdapter.LabeledItemBaseViewHolder
+            implements Marker.OnMarkerDragListener {
+        private final MapView mMapView;
+        private final int mRowHeightPixels;
+        private Item mBoundItem;
+        private boolean mStarted;
+
+        public OsmViewHolder(LayoutInflater inflater, ViewGroup parent,
+                Connection connection, WidgetAdapter.ColorMapper colorMapper) {
+            super(inflater, parent, R.layout.openhabwidgetlist_mapitem, connection, colorMapper);
+            mMapView = itemView.findViewById(R.id.mapview);
+            mMapView.setTileSource(TileSourceFactory.MAPNIK);
+
+            mMapView.setVerticalMapRepetitionEnabled(false);
+            mMapView.getOverlays().add(new CopyrightOverlay(itemView.getContext()));
+            mMapView.setBuiltInZoomControls(false);
+            mMapView.setMultiTouchControls(false);
+            mMapView.getOverlays().add(new MapEventsOverlay(new MapEventsReceiver() {
+                @Override
+                public boolean singleTapConfirmedHelper(GeoPoint p) {
+                    openPopup();
+                    return true;
+                }
+
+                @Override
+                public boolean longPressHelper(GeoPoint p) {
+                    return false;
+                }
+            }));
+
+            new Handler().post(() -> applyPositionAndLabel(mMapView, 15.0f, false, false));
```

Comment:
Is the handler really needed? If so, should probably make it a member variable considering it's used multiple times.

Reply:
If I don't use the Handler here, the markers are applied before the MapView is inflated, leading to runtime crashes.

Reply:
I can not reproduce that using the emulator. How does the crash trace look like for you?

Reply:
```
05-10 22:53:23.356 org.openhab.habdroid.beta D/MapViewHelper: North 52.516998, south 52.506340, west 10.388490, east 15.388490
05-10 22:53:23.357 org.openhab.habdroid.beta D/AndroidRuntime: Shutting down VM
05-10 22:53:23.361 org.openhab.habdroid.beta E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid.beta, PID: 7685
    java.lang.IllegalArgumentException: north must be in [-85.05112877980658,85.05112877980658]
        at org.osmdroid.util.BoundingBox.set(BoundingBox.java:67)
        at org.osmdroid.views.Projection.refresh(Projection.java:680)
        at org.osmdroid.views.Projection.<init>(Projection.java:99)
        at org.osmdroid.views.Projection.<init>(Projection.java:110)
        at org.osmdroid.views.MapView.zoomToBoundingBox(MapView.java:556)
        at org.osmdroid.views.MapView.zoomToBoundingBox(MapView.java:582)
        at org.openhab.habdroid.ui.MapViewHelper$OsmViewHolder.applyPositionAndLabel(MapViewHelper.java:207)
```

Opened an issue here: https://github.com/osmdroid/osmdroid/issues/1339#issuecomment-491487934

This only occurs if you have more than one markers and remove all the Handlers and the try/catch below.

Reply:
Ah, OK, only tried with a single item. Thanks for that link giving context; let's see what they'll say. The `Handler` seems like a workaround for an issue on their side ... and if we want to keep it, we might rather attach a `ViewTreeObserver.OnGlobalLayoutListener` to see whether the map is already laid out.

Reply:
(And BTW, you're using incorrect terminology there: the view is already `inflated` (XML description was used to create a view hierarchy), it's just not laid out yet)